src/rebase.rs

```diff
@@ -0,0 +1,377 @@
+                                                      })
+                                                 .unwrap_or(0 as *const raw::git_signature)));
+        }
+        Ok(())
+    }
+    /// Gets the index produced by the last operation.
+    /// This is useful for resolving conflicts in an in-memory rebase before committing them.
+    pub fn inmemory_index(&self) -> Result<Index, Error> {
+        let mut ret = 0 as *mut raw::git_index;
+        unsafe {
+            try_call!(raw::git_rebase_inmemory_index(&mut ret, self.raw));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Performs the next rebase operation and returns the `RebaseOperation` about it.
+    ///
+    /// This is the fundamental operation that `operation_iter` relies upon
+    pub fn next<'rebase>(&self) -> Result<RebaseOperation<'rebase, 'repo>, Error> {
+        let mut ret = 0 as *mut raw::git_rebase_operation;
+        unsafe {
+            try_call!(raw::git_rebase_next(&mut ret, self.raw));
+            Ok(Binding::from_raw(ret))
+        }
+    }
+
+    /// Iterator of rebase operations
+    pub fn operation_iter(&self) -> RebaseOperations {
+        RebaseOperations {
+            rebase: self,
+            count: 0,
+        }
+    }
+
+    /// Gets the rebase operation specified by the given index.
+    pub fn operation_at_index(&self, index: usize) -> Option<RebaseOperation> {
+        unsafe {
+            let ptr = raw::git_rebase_operation_byindex(self.raw, index as size_t);
+            if ptr.is_null() {
+                None
+            } else {
+                Some(Binding::from_raw(ptr))
+            }
+        }
+    }
+
+    /// Gets the index of the rebase operation that is currently being applied.
+    /// If the first operation has not yet been applied it returns `None`.
+    pub fn current_operation_index(&self) -> Option<usize> {
+        unsafe {
+            let index = raw::git_rebase_operation_current(self.raw);
+            if index == raw::GIT_REBASE_NO_OPERATION {
+                None
+            } else {
+                Some(index as usize)
+            }
+        }
+    }
+
+    /// Convenience function to get the operation at the current index.
+    /// Returns none if the first operation has not yet been applied.
+    pub fn current_operation(&self) -> Option<RebaseOperation> {
+        if let Some(index) = self.current_operation_index() {
+            self.operation_at_index(index)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'rebase, 'repo: 'rebase> Iterator for RebaseOperations<'rebase, 'repo> {
+    type Item = Result<RebaseOperation<'rebase, 'repo>, Error>;
+    fn next(&mut self) -> Option<Result<RebaseOperation<'rebase, 'repo>, Error>> {
+        if self.count >= self.rebase.operation_count() {
+            None
+        } else {
+            self.count += 1;
+            Some(self.rebase.next())
```

Comment:
Hm this means that calling `.operations_iter()` will change over time, right? Should this perhaps use `operation_at_index`?

Reply:
tl;dr: I've changed this to `self.rebase.current_operation_index().map(|_| self.rebase.next())` and removed the `count` field.

This function is the only place where `self.count` is ever accessed or mutated. It appears to just be a way of keeping track of where we are in the rebase to make sure we don't call the underlying `git_rebase_operation_next` function incorrectly. I believe that git's internal `current_operation_index` function will be re-entrant, where `count` will fail if you try to use this iterator.

Based on [the docs for `git_rebase_next`](https://libgit2.github.com/libgit2/#HEAD/group/rebase/git_rebase_next) I believe that we need to call it in order to actually make progress in the rebase. [`git_rebase_operation_byindex`](https://libgit2.github.com/libgit2/#HEAD/group/rebase/git_rebase_operation_byindex) doesn't say anything that suggests that it interacts with any sort of internal state machine. So I don't think it's fine to swap to just using `operation_at_index`.